### PR TITLE
Convert JSDOM to use callback functions

### PR DESCRIPTION
**Pull request by ExE-Boss**
## Depends on:

- [x] <https://github.com/jsdom/webidl2js/pull/194> (merged)
- [x] indirectly on <https://github.com/jsdom/jsdom/pull/2958> (as a result of <https://github.com/jsdom/webidl2js/pull/206> getting merged before <https://github.com/jsdom/webidl2js/pull/194>)

---

<details>
<summary><strong>Benchmark results</strong></summary>

| benchmark                                                                                    | `master`                                                                                        | `feat/use‑callback‑functions`
| ---------                                                                                    | --------                                                                                        | --------------------------------------
| dom/compare-document-position/compare&nbsp;ancestor                                          | jsdom  ×&nbsp;31,761&nbsp;ops/sec ±2.80%&nbsp;(10&nbsp;runs&nbsp;sampled)                       | jsdom  ×&nbsp;30,750&nbsp;ops/sec ±5.61%&nbsp;(9&nbsp;runs&nbsp;sampled)
| dom/compare-document-position/compare&nbsp;descendant                                        | jsdom  ×&nbsp;32,439&nbsp;ops/sec ±1.92%&nbsp;(10&nbsp;runs&nbsp;sampled)                       | jsdom  ×&nbsp;30,535&nbsp;ops/sec ±3.14%&nbsp;(9&nbsp;runs&nbsp;sampled)
| dom/compare-document-position/compare&nbsp;siblings                                          | jsdom  ×&nbsp;1,224,691&nbsp;ops/sec ±13.62%&nbsp;(10&nbsp;runs&nbsp;sampled)                   | jsdom  ×&nbsp;1,352,267&nbsp;ops/sec ±2.21%&nbsp;(10&nbsp;runs&nbsp;sampled)
| dom/construction/createComment                                                               | jsdom  ×&nbsp;1,243,110&nbsp;ops/sec ±0.58%&nbsp;(97&nbsp;runs&nbsp;sampled)                    | jsdom  ×&nbsp;1,219,938&nbsp;ops/sec ±0.31%&nbsp;(98&nbsp;runs&nbsp;sampled)
| dom/construction/createDocumentFragment                                                      | jsdom  ×&nbsp;1,132,790&nbsp;ops/sec ±1.83%&nbsp;(96&nbsp;runs&nbsp;sampled)                    | jsdom  ×&nbsp;1,165,658&nbsp;ops/sec ±0.45%&nbsp;(97&nbsp;runs&nbsp;sampled)
| dom/construction/createElement                                                               | jsdom  ×&nbsp;295,458&nbsp;ops/sec ±0.43%&nbsp;(96&nbsp;runs&nbsp;sampled)                      | jsdom  ×&nbsp;293,850&nbsp;ops/sec ±0.27%&nbsp;(100&nbsp;runs&nbsp;sampled)
| dom/construction/createEvent                                                                 | jsdom  ×&nbsp;204,673&nbsp;ops/sec ±0.97%&nbsp;(94&nbsp;runs&nbsp;sampled)                      | jsdom  ×&nbsp;208,309&nbsp;ops/sec ±0.48%&nbsp;(98&nbsp;runs&nbsp;sampled)
| dom/construction/createNodeIterator                                                          | jsdom  ×&nbsp;1,257,055&nbsp;ops/sec ±0.67%&nbsp;(97&nbsp;runs&nbsp;sampled)                    | jsdom  ×&nbsp;1,233,311&nbsp;ops/sec ±0.31%&nbsp;(92&nbsp;runs&nbsp;sampled)
| dom/construction/createProcessingInstruction                                                 | jsdom  ×&nbsp;625,897&nbsp;ops/sec ±0.54%&nbsp;(94&nbsp;runs&nbsp;sampled)                      | jsdom  ×&nbsp;625,065&nbsp;ops/sec ±1.97%&nbsp;(94&nbsp;runs&nbsp;sampled)
| dom/construction/createTextNode                                                              | jsdom  ×&nbsp;1,014,577&nbsp;ops/sec ±2.88%&nbsp;(91&nbsp;runs&nbsp;sampled)                    | jsdom  ×&nbsp;1,036,870&nbsp;ops/sec ±0.40%&nbsp;(96&nbsp;runs&nbsp;sampled)
| dom/inner-html/tables                                                                        | jsdom  ×&nbsp;0.33&nbsp;ops/sec ±5.08%&nbsp;(5&nbsp;runs&nbsp;sampled)                          | jsdom  ×&nbsp;0.32&nbsp;ops/sec ±4.12%&nbsp;(5&nbsp;runs&nbsp;sampled)
| dom/named-properties/setAttribute(): Remove&nbsp;a&nbsp;named&nbsp;property from&nbsp;window | jsdom  ×&nbsp;3,033&nbsp;ops/sec ±1.27%&nbsp;(58&nbsp;runs&nbsp;sampled)                        | jsdom  ×&nbsp;3,094&nbsp;ops/sec ±5.09%&nbsp;(54&nbsp;runs&nbsp;sampled)
| dom/tree-modification/appendChild: many&nbsp;parents                                         | jsdom  ×&nbsp;24.64&nbsp;ops/sec ±0.57%&nbsp;(43&nbsp;runs&nbsp;sampled)                        | jsdom  ×&nbsp;27.23&nbsp;ops/sec ±0.60%&nbsp;(47&nbsp;runs&nbsp;sampled)
| dom/tree-modification/appendChild: many&nbsp;siblings                                        | jsdom  ×&nbsp;243,263&nbsp;ops/sec ±11.94%&nbsp;(18&nbsp;runs&nbsp;sampled)                     | jsdom  ×&nbsp;240,553&nbsp;ops/sec ±12.88%&nbsp;(18&nbsp;runs&nbsp;sampled)
| dom/tree-modification/appendChild: no&nbsp;siblings                                          | jsdom  ×&nbsp;234,538&nbsp;ops/sec ±5.76%&nbsp;(26&nbsp;runs&nbsp;sampled)                      | jsdom  ×&nbsp;233,948&nbsp;ops/sec ±6.95%&nbsp;(25&nbsp;runs&nbsp;sampled)
| dom/tree-modification/insertBefore: many&nbsp;siblings                                       | jsdom  ×&nbsp;206,035&nbsp;ops/sec ±28.30%&nbsp;(18&nbsp;runs&nbsp;sampled)                     | jsdom  ×&nbsp;222,148&nbsp;ops/sec ±15.60%&nbsp;(16&nbsp;runs&nbsp;sampled)
| dom/tree-modification/removeChild: many&nbsp;parents                                         | jsdom  ×&nbsp;126&nbsp;ops/sec ±2.90%&nbsp;(63&nbsp;runs&nbsp;sampled)                          | jsdom  ×&nbsp;134&nbsp;ops/sec ±0.57%&nbsp;(62&nbsp;runs&nbsp;sampled)
| dom/tree-modification/removeChild: many&nbsp;siblings                                        | jsdom  ×&nbsp;150,933&nbsp;ops/sec ±6.34%&nbsp;(39&nbsp;runs&nbsp;sampled)                      | jsdom  ×&nbsp;160,629&nbsp;ops/sec ±6.72%&nbsp;(42&nbsp;runs&nbsp;sampled)
| dom/tree-modification/removeChild: no&nbsp;siblings                                          | jsdom  ×&nbsp;143,544&nbsp;ops/sec ±5.47%&nbsp;(30&nbsp;runs&nbsp;sampled)                      | jsdom  ×&nbsp;138,649&nbsp;ops/sec ±8.73%&nbsp;(29&nbsp;runs&nbsp;sampled)
| html/parsing/text                                                                            | jsdom  ×&nbsp;15.80&nbsp;ops/sec ±1.44%&nbsp;(44&nbsp;runs&nbsp;sampled)                        | jsdom  ×&nbsp;15.49&nbsp;ops/sec ±1.27%&nbsp;(43&nbsp;runs&nbsp;sampled)
| jsdom/api/new&nbsp;JSDOM() defaults                                                          | &lt;Test&nbsp;#&#x2060;10&gt; ×&nbsp;278&nbsp;ops/sec ±3.24%&nbsp;(81&nbsp;runs&nbsp;sampled)   | &lt;Test&nbsp;#&#x2060;10&gt; ×&nbsp;282&nbsp;ops/sec ±2.23%&nbsp;(84&nbsp;runs&nbsp;sampled)
| jsdom/api/new&nbsp;JSDOM() with&nbsp;many&nbsp;elements                                      | &lt;Test&nbsp;#&#x2060;11&gt; ×&nbsp;29.25&nbsp;ops/sec ±5.29%&nbsp;(54&nbsp;runs&nbsp;sampled) | &lt;Test&nbsp;#&#x2060;11&gt; ×&nbsp;29.05&nbsp;ops/sec ±8.35%&nbsp;(53&nbsp;runs&nbsp;sampled)
</details>

### CODE REVIEW

File: lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js [hunk 1]
```diff
@@ -25,7 +26,7 @@ function convertToSequenceDOMString(obj) {
     throw new TypeError("Invalid Sequence");
   }
 
-  return Array.from(obj).map(webIDLConversions.DOMString);
+  return Array.from(obj, v => webIDLConversions.DOMString(v));
```
Comment (domenic) [inline lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js:29]: Please don't make unrelated changes in the PR. You can file issues asking if they are OK but it is not good to do so unsolicited.

File: lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js [hunk 2]
```diff
@@ -51,7 +52,7 @@ function isConstructor(value) {
   }
 }
 
-// https://html.spec.whatwg.org/#customelementregistry
+// https://html.spec.whatwg.org/multipage/custom-elements.html#customelementregistry
```
Comment (domenic) [inline lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js:55]: Please don't make unrelated changes in the PR. You can file issues asking if they are OK but it is not good to do so unsolicited.

File: lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js [hunk 3]
```diff
@@ -61,9 +62,10 @@ class CustomElementRegistryImpl {
     this._globalObject = globalObject;
   }
 
-  // https://html.spec.whatwg.org/#dom-customelementregistry-define
-  define(name, ctor, options) {
+  // https://html.spec.whatwg.org/multipage/custom-elements.html#dom-customelementregistry-define
```
Comment (domenic) [inline lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js:65]: Please don't make unrelated changes in the PR. You can file issues asking if they are OK but it is not good to do so unsolicited.

File: lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js [hunk 4]
```diff
@@ -61,9 +62,10 @@ class CustomElementRegistryImpl {
...
+  define(name, ctorWrapper, options) {
```
Comment (domenic) [inline lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js:66]: Please don't make unrelated changes in the PR. You can file issues asking if they are OK but it is not good to do so unsolicited.

File: lib/jsdom/living/helpers/create-event-accessor.js [hunk 5]
```diff
@@ -77,114 +74,109 @@ exports.setupForSimpleEventAccessors = (prototype, events) => {
...
+
+    target._setEventHandlerFor(event, fn);
+  }
+
+  return target._getEventHandlerFor(event);
+};
+
+// https://html.spec.whatwg.org/multipage/webappapis.html#event-handler-idl-attributes
+// TODO: Consider replacing this with `[ReflectEvent]`
+exports.createEventAccessor = function createEventAccessor(obj, event) {
```
Comment (domenic) [inline lib/jsdom/living/helpers/create-event-accessor.js:171]:
```suggestion
exports.createEventAccessor = (obj, event) => {
```

File: lib/jsdom/living/nodes/GlobalEventHandlers.webidl [hunk 6]
```diff
@@ -1,4 +1,4 @@
-// https://html.spec.whatwg.org/#globaleventhandlers
+// https://html.spec.whatwg.org/multipage/webappapis.html#globaleventhandlers
```
Comment (domenic) [inline lib/jsdom/living/nodes/GlobalEventHandlers.webidl:1]: Please don't make unrelated changes in the PR. You can file issues asking if they are OK but it is not good to do so unsolicited.

File: lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js (hunk 4 above)
Comment (ExE-Boss) [reply lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js:66]: If I didn’t do this, then it would have issues later, because the `CustomElementRegistry` implementation works with the `objectReference`, rather than with the `invokeTheCallbackFunction` wrapper.

Comment (domenic) [reply lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js:66]: Renaming is not necessary.

**Comment by TimothyGu** [comment]
> Can you check the performance impact of this approach using the benchmark suite?

I'd be interested in this as well, especially the memory usage.

**Comment by ExE-Boss** [comment]
I’ve now performed the benchmark.

File: lib/jsdom/browser/Window.js [hunk 7]
```diff
@@ -8,6 +8,9 @@ const { installInterfaces } = require("../living/interfaces");
 const { define, mixin } = require("../utils");
 const Element = require("../living/generated/Element");
 const EventTarget = require("../living/generated/EventTarget");
+const EventHandlerNonNull = require("../living/generated/EventHandlerNonNull.js");
+const OnBeforeUnloadEventHandlerNonNull = require("../living/generated/OnBeforeUnloadEventHandlerNonNull.js");
+const OnErrorEventHandlerNonNull = require("../living/generated/OnErrorEventHandlerNonNull.js");
```
Comment (domenic) [inline lib/jsdom/browser/Window.js:13]: Nit: although I like the .js, it'd be better to be consistent with the nearby requires, which don't have it.

I'd support a separate PR that integrates https://www.npmjs.com/package/eslint-plugin-node and enforces extensions throughout the project.

File: lib/jsdom/browser/Window.js [hunk 8]
```diff
@@ -97,6 +154,62 @@ function setupWindow(windowInstance, { runScripts }) {
...
   windowInstance._initGlobalEvents();
 
+  Object.defineProperty(windowInstance, "onbeforeunload", {
+    configurable: true,
+    enumerable: true,
+    get() {
+      return idlUtils.tryWrapperForImpl(getCurrentEventHandlerValue(this, "beforeunload"));
+    },
+    set(V) {
+      if (!idlUtils.isObject(V)) {
```
Comment (domenic) [inline lib/jsdom/browser/Window.js:164]: Why is this necessary? Shouldn't OnBeforeUnloadEventHandlerNonNull handle the [LegacyTreatNonObjectAsNull] ness?

File: lib/jsdom/living/events/EventHandler.webidl [hunk 9]
```diff
@@ -0,0 +1,20 @@
+// https://html.spec.whatwg.org/multipage/webappapis.html#eventhandlernonnull
```
Comment (domenic) [inline lib/jsdom/living/events/EventHandler.webidl:1]: Hmm, maybe name this file event-handlers.webidl since it contains more than just EventHandler and its supporting definitions?

File: lib/jsdom/browser/Window.js (hunk 8 above)
Comment (ExE-Boss) [reply lib/jsdom/browser/Window.js:164]: The non‑`object` parameter is handled as part of <https://heycam.github.io/webidl/#es-nullable-type>.

Comment (domenic) [reply lib/jsdom/browser/Window.js:164]: Sorry, I don't understand. Can you explain more why `OnBeforeUnloadEventHandlerNonNull.convert()` fails to handle non-objects?

Comment (ExE-Boss) [reply lib/jsdom/browser/Window.js:164]: This is because `OnBeforeUnloadEventHandlerNonNull.convert()` [always returns a `invokeTheCallbackFunction` operation](https://heycam.github.io/webidl/#es-callback-function), because it’s a non‑nullable type.

In order for the conversion to return `null`, it has to be wrapped in a [nullable type conversion](https://heycam.github.io/webidl/#es-nullable-type), where the first step is:

> 1. If [Type](https://tc39.github.io/ecma262/#sec-ecmascript-data-types-and-values)(<var>V</var>) is not Object, and the conversion to an IDL value is being performed due to <var>V</var> being assigned to an [attribute](https://heycam.github.io/webidl/#dfn-attribute) whose type is a [nullable](https://heycam.github.io/webidl/#dfn-nullable-type) [callback function](https://heycam.github.io/webidl/#dfn-callback-function) that is annotated with [[`LegacyTreatNonObjectAsNull`](https://heycam.github.io/webidl/#LegacyTreatNonObjectAsNull)], then return the IDL [nullable type](https://heycam.github.io/webidl/#dfn-nullable-type) <var>T</var>? value **null**.

Comment (domenic) [reply lib/jsdom/browser/Window.js:164]: Then is there a spec bug? https://html.spec.whatwg.org/#event-handler-idl-attributes setter algorithm step 3 says "If the given value is null". But you're saying that it will never be null after passing through the IDL layer?

Comment (domenic) [reply lib/jsdom/browser/Window.js:164]: Oh, I see, I was confused.

OK, so my question is, why can't we use `OnBeforeUnloadEventHandler.convert(V, ...)`?

Comment (TimothyGu) [reply lib/jsdom/browser/Window.js:164]: `OnBeforeUnloadEventHandler`, as a typedef, does not generate conversion files. We could make them generate files, but a majority of them don't get used externally. An webidl2js-specific extended attribute on the typedefs to force-generate the files may be appropriate.

Comment (domenic) [reply lib/jsdom/browser/Window.js:164]: Hmm, I see. So if Window was converted to Web IDL, then it would be able to use the typedef, but we can't use it manually?

Comment (TimothyGu) [reply lib/jsdom/browser/Window.js:164]: Yes, at least right now.

However, if I were to have to pick the easier of converting `Window` entirely to Web IDL and making typedefs usable outside of Web IDL, I would go for the latter every time :')

Comment (domenic) [reply lib/jsdom/browser/Window.js:164]: OK, great, I'm satisfied then. Although I'd like a comment to explain this; I can push a commit that adds that.

I feel like we're relatively close to converting Window to use Web IDL...

**Comment by TimothyGu** [comment]
It looks like this branch needs some rebasing. I'm wary of the green "squash & merge" button when the branch has merge commits.

File: lib/jsdom/browser/Window.js (hunk 7 above)
Comment (domenic) [reply lib/jsdom/browser/Window.js:13]: This nit still remains

Comment (ExE-Boss) [reply lib/jsdom/browser/Window.js]: It’s [eslint‑plugin‑import](https://www.npmjs.com/package/eslint-plugin-import) that enforces extensions in `import`s and `require(…)`s.

**Comment by domenic** [comment]
@ExE-Boss I rolled to webidl2js 16.2.0 in the first commit as well. I'm unsure whether that, or other changes to webidl2.js in the meantime, has caused the breakage with `composed` in `getRootNode`. It seems likely to be a general problem with dictionaries, likely due to the change in how their default syntax works. The generated code does not provide proper defaults for the dictionary when no arguments are passed...

**Comment by ExE-Boss** [comment]
@domenic
> I rolled to webidl2js 16.2.0 in the first commit as well. I'm unsure whether that, or other changes to webidl2.js in the meantime, has caused the breakage with `composed` in `getRootNode`. It seems likely to be a general problem with dictionaries, likely due to the change in how their default syntax works. The generated code does not provide proper defaults for the dictionary when no arguments are passed...

I am unable to reproduce that on my end.

**Comment by domenic** [comment]
I think I forgot to run `yarn pretest`. It should be good now.
